File: library/src/main/java/cc/shinichi/library/glide/progress/ProgressManager.java
package cc.shinichi.library.glide.progress;

import android.text.TextUtils;

import java.io.IOException;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/**
 * @author 工藤
 * @email qinglingou@gmail.com
 */
public class ProgressManager {

    private static final Map<String, OnProgressListener> listenersMap =
            Collections.synchronizedMap(new HashMap<String, OnProgressListener>());

    private static final ProgressResponseBody.InternalProgressListener LISTENER =
            new ProgressResponseBody.InternalProgressListener() {
                @Override
                public void onProgress(String url, long bytesRead, long totalBytes) {
                    OnProgressListener onProgressListener = getProgressListener(url);
                    if (onProgressListener != null) {
                        int percentage = (int) ((bytesRead * 1f / totalBytes) * 100f);
                        boolean isComplete = percentage >= 100;
                        onProgressListener.onProgress(url, isComplete, percentage, bytesRead, totalBytes);
                        if (isComplete) {
                            removeListener(url);
                        }
                    }
                }
            };

    private ProgressManager() {

    }

    /**
     * 下载监听
     * @return
     */
    public static Interceptor getNetWorkInterceptor(){
       return new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Request request = chain.request();
                Response response = chain.proceed(request);
                return response.newBuilder()
                        .body(new ProgressResponseBody(request.url().toString(), LISTENER, response.body()))
                        .build();
            }
        };
    }

    public static void addListener(String url, OnProgressListener listener) {
        if (!TextUtils.isEmpty(url) && listener != null) {
            listenersMap.put(url, listener);
            listener.onProgress(url, false, 1, 0, 0);
        }
    }

    public static void removeListener(String url) {
        if (!TextUtils.isEmpty(url)) {
            listenersMap.remove(url);
        }
    }

    public static OnProgressListener getProgressListener(String url) {
        if (TextUtils.isEmpty(url) || listenersMap == null || listenersMap.size() == 0) {
            return null;
        }

        OnProgressListener listenerWeakReference = listenersMap.get(url);
        return listenerWeakReference;
    }
}

File: library/src/main/java/cc/shinichi/library/glide/progress/ProgressManager.kt
package cc.shinichi.library.glide.progress

import android.text.TextUtils
import cc.shinichi.library.ImagePreview
import cc.shinichi.library.glide.SSLSocketClient
import cc.shinichi.library.glide.progress.ProgressResponseBody.InternalProgressListener
import okhttp3.Headers
import okhttp3.OkHttpClient
import java.util.Collections
import java.util.concurrent.TimeUnit

/**
 * @author 工藤
 * @email qinglingou@gmail.com
 */
object ProgressManager {

    private val listenersMap = Collections.synchronizedMap(HashMap<String, OnProgressListener>())

    private val LISTENER = object : InternalProgressListener {
        override fun onProgress(url: String?, bytesRead: Long, totalBytes: Long) {
            val percentage = (bytesRead * 1f / totalBytes * 100f).toInt()
            val isComplete = percentage >= 100
            listenersMap.let {
                for (listener in it.values) {
                    listener.onProgress(url, isComplete, percentage, bytesRead, totalBytes)
                }
            }
            if (isComplete) {
                removeListener(url)
            }
        }
    }

    @JvmStatic
    val okHttpClient: OkHttpClient
        get() {
            val builder = OkHttpClient.Builder()
            builder.addNetworkInterceptor { chain ->
                val request = chain.request()
                val response = chain.proceed(request)
                val headersBuilder = Headers.Builder()
                val url = request.url().toString()
                // 如果url中包含任意一个关键字，就添加监听
                val hostKeywordList = ImagePreview.instance.hostKeywordList
                var needAddHeader = false
                if (hostKeywordList.isNullOrEmpty()) {
                    needAddHeader = false
                } else {
                    for (item in hostKeywordList) {
                        if (url.contains(item)) {
                            needAddHeader = true
                            break
                        }
                    }
                }
                if (needAddHeader) {
                    // 通过set，替换原本的header
                    ImagePreview.instance.headers?.forEach {
                        headersBuilder.set(it.key, it.value)
                    }
                }
                response.newBuilder()
                    .headers(headersBuilder.build())
                    .body(
                        response.body()
                            ?.let { ProgressResponseBody(request.url().toString(), LISTENER, it) })
                    .build()
            }
                .sslSocketFactory(
                    SSLSocketClient.sSLSocketFactory,
                    SSLSocketClient.geX509tTrustManager()
                )
                .hostnameVerifier(SSLSocketClient.hostnameVerifier)
            builder.connectTimeout(30, TimeUnit.SECONDS)
            builder.writeTimeout(30, TimeUnit.SECONDS)
            builder.readTimeout(30, TimeUnit.SECONDS)
            return builder.build()
        }

    @JvmStatic
    fun addListener(url: String?, listener: OnProgressListener?) {
        if (!TextUtils.isEmpty(url)) {
            listenersMap[url] = listener
            listener?.onProgress(url, false, 1, 0, 0)
        }
    }

    private fun removeListener(url: String?) {
        if (!TextUtils.isEmpty(url)) {
            listenersMap.remove(url)
        }
    }
}